File: Algorithms/Sorting/Quick Sort/QuickSort.java
// package Sorting;

import java.util.Scanner;
import java.io.File;
import java.io.FileNotFoundException;

public class QuickSort{

    private void sort(int[] array, int start, int end){

        if (start < end){
            int mid = partition(array, start, end);
            sort(array, start, mid - 1);
            sort(array, mid + 1, end);
        }
    }

    private int partition(int[] array, int start, int end){

        int pivot = array[end];
        int i = start - 1;

        for (int k = start; k < end; k++){
            if(array[k] <= pivot){
                i += 1;
                int temp = array[k];
                array[k] = array[i];
                array[i] = temp;
            }
        }
        int temp = array[end];
        array[end] = array[i + 1];
        array[i + 1] = temp;

        return i + 1;
    }


    public static void main(String[] args) throws FileNotFoundException{
        File file = new File("QuickSortInput.txt");
        Scanner sc = new Scanner(file);

        int n = sc.nextInt();
        int[] array = new int[n];

        for(int i = 0; i < n; i++){
            array[i] = sc.nextInt();
        }

        QuickSort qs = new QuickSort();
        qs.sort(array, 0, n - 1);

        for (int i: array){
            System.out.print(i + " ");
        }
        sc.close();
    }
}

File: Algorithms/Sorting/Quick Sort/QuickSort.cpp
#include <iostream>
#include <fstream>

using namespace std;

int partition(int* Array, int start, int end);
void QuickSort(int* Array, int start, int end);


int main(){

    fstream input;
    input.open("QuickSortInput.txt");

    int n; 
    input >> n;

    int *array = new int[n];

    for (int i = 0; i < n; i++){
        input >> array[i];
    }

    QuickSort(array, 0, n - 1);

    for (int i = 0; i < n; i++){
        cout << array[i] << " ";
    }

    input.close();
    delete[] array;
    return 0;
}

void QuickSort(int* Array, int start, int end){
    if (start < end){
        int mid = partition(Array, start, end);
        QuickSort(Array, start, mid - 1);
        QuickSort(Array, mid + 1, end);
    }
}

int partition(int *Array, int start, int end){

    int pivot = Array[end];
    int i = start - 1;

    for (int j = start; j < end; j++){

        if (Array[j] <= pivot){
            ++i;
            int temp = Array[j];
            Array[j] = Array[i];
            Array[i] = temp;
        }
    }
    int temp = Array[end];
    Array[end] = Array[i + 1];
    Array[i + 1] = temp;
    
    return i + 1;
}